### DISCUSSION TITLE
(ANSWERED) Calculate phase angle for a satellite

### DISCUSSION THREAD

**Discussion by Space-A** [skyfielders/python-skyfield#607] [Q&A] [answered]
Hi!

I'm working on a code to calculate the apparent magnitude of a satellite, and for this I need to calculate the **phase angle** between the observer location, satellite and sun. Is there an obvious way to do this using Skyfield? 

I have seen that there is a `is_sunlit ` function which I suspect uses the phase angle, but the function outputs True/False and not the angle itself. I have also seen the `separation_from(another_icrf) ` function which I think also can be useful, but I'm still unsure about exactly how to use it in this specific case.

**Comment by JoshPaterson** [comment, marked as answer]
Hi! If I understand correctly you are looking for the angle between the sun and the earth as observed by a satellite. Here is an example which does that for the International Space Station using the `separation_from` method. I hope it helps!

```
 from skyfield.api import load

ts = load.timescale()
eph = load('de421.bsp')
earth = eph['earth']
sun = eph['sun']

stations_url = 'http://celestrak.com/NORAD/elements/stations.txt'
satellites = load.tle_file(stations_url)

by_name = {sat.name: sat for sat in satellites}
ISS = earth + by_name['ISS (ZARYA)']

t = ts.utc(2021, 6, 1)

sun_position = ISS.at(t).observe(sun)
earth_position = ISS.at(t).observe(earth)

phase_angle = sun_position.separation_from(earth_position)

print(phase_angle)
```
Outputs:
```
42deg 56' 16.6"
```

**Reply by Space-A** [reply]
Thank you very much! I managed to calculate it differently, but your way is much quicker and straight forward.

**Reply by yoannq** [reply]
hi 
additionnal question , if we need to compute phase angle between specific_location , sat , sun we just need to add location to earth as follow ? is it correct ?
best regards

```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sat Jun 26 09:02:29 2021

@author: yoann
"""

import datetime as dt
from pytz import timezone
from skyfield import almanac
from skyfield.api import EarthSatellite,wgs84, load



### iss
line1 = '1 25544U 98067A   21177.11584491  .00002721  00000-0  58043-4 0  9997'
line2 = '2 25544  51.6453 298.5670 0002273 106.5370 241.5039 15.48733308289970'

ts = load.timescale()
t=ts.utc(2021, 6, 26, 10, 30)
planets = load('de421.bsp')
earth = planets['earth']
sun = planets['sun']
### maison
obs = wgs84.latlon(45.186089, -1.057202)
t.utc_iso()
satellite = EarthSatellite(line1, line2, name='ISS (ZARYA)')

ssb_obs = earth + obs
ssb_satellite = earth + satellite
topocentric_sat_obs = ssb_satellite.at(t).observe(ssb_obs).apparent()
topocentric_sat_sun = ssb_satellite.at(t).observe(sun).apparent()
phase_angle = topocentric_sat_obs.separation_from(topocentric_sat_sun)
print('Angle between the two positions in the sky:')
print('{}'.format(phase_angle.degrees))
```

**Reply by brandon-rhodes** [reply]
@yoannq —

1. I have edited your question to put a line with three backticks on top of and on bottom of your code, which hopefully will format it so that folks can read it more easily. Double-check and see whether that changed the code at all.
2. Your code does seem to compute, from the satellite's point of view, the angle between the observer in its sky and the Sun in its Sky!

**Comment by Madalinasst** [comment]
Hello everyone! I am interested in the same thing, to compute the phase angle between the observer location, satellite and sun, but as seen from the Earth! From what  I understand from your comments (@brandon-rhodes), the above code is used to compute the phase angle, but from the point of view of the satellite. Can you help me with the code that computes the phase angle between the observer-satellite-sun as seen from the observer's point of view?

Thank you very much!
Regards,
Madalina

**Reply by brandon-rhodes** [reply]
@MadalinaTrelia — Try reading through the comments of the accepted answer. @yoannq asked what I think is the same question, and supplied sample code that I think did the calculation successfully. Try it out yourself on a few satellites and see if the answers look roughly like you expect, and let us know if the technique worked!

**Comment by Madalinasst** [comment]
Sure, I did that and it worked very nice, I have obtained the angle very easy, but  I was not sure if the angle was obtained for the observer's or the satellite's point of view. Thanks a lot for the code (and answer), it's very helpful for those of us that are not that skilled in programming!

**Reply by brandon-rhodes** [reply]
I might have misunderstood your question: when you say “observer-satellite-sun as seen from the observer's point of view?” do you perhaps mean “satellite-observer-sun”, with the observer as the vertex of the angle rather than the satellite at the vertex? In that case you can simply `.observe()` both the satellite and the sun from the observer, and then compute the `.separation()` between the two angles. I think that angle is called the "elongation", in which case you can see an example here:

https://rhodesmill.org/skyfield/examples.html#when-is-venus-at-its-greatest-east-and-west-elongations-from-the-sun

**Comment by Madalinasst** [comment]
Sorry for the misunderstanding. The right angle is observer-satellite-sun with the satellite as the vertex of the angle, as you said. Thanks again!

**Comment by izzatzubir** [comment]
Is there any difference between using the separation_from() method vs using phase_angle() method?
